protocol/messenger_share_urls_test.go

```diff
@@ -0,0 +1,273 @@
+)
+
+func TestMessengerShareUrlsSuite(t *testing.T) {
+	suite.Run(t, new(MessengerShareUrlsSuite))
+}
+
+type MessengerShareUrlsSuite struct {
+	suite.Suite
+	m          *Messenger        // main instance of Messenger
+	privateKey *ecdsa.PrivateKey // private key for the main instance of Messenger
+	// If one wants to send messages between different instances of Messenger,
+	// a single waku service should be shared.
+	shh    types.Waku
+	logger *zap.Logger
+}
+
+func (s *MessengerShareUrlsSuite) SetupTest() {
+	s.logger = tt.MustCreateTestLogger()
+
+	config := waku.DefaultConfig
+	config.MinimumAcceptedPoW = 0
+	shh := waku.New(&config, s.logger)
+	s.shh = gethbridge.NewGethWakuWrapper(shh)
+	s.Require().NoError(shh.Start())
+
+	s.m = s.newMessenger()
+	s.privateKey = s.m.identity
+	_, err := s.m.Start()
+	s.Require().NoError(err)
+}
+
+func (s *MessengerShareUrlsSuite) TearDownTest() {
+	s.Require().NoError(s.m.Shutdown())
+}
+
+func (s *MessengerShareUrlsSuite) newMessenger() *Messenger {
+	privateKey, err := crypto.GenerateKey()
+	s.Require().NoError(err)
+
+	messenger, err := newMessengerWithKey(s.shh, privateKey, s.logger, nil)
+	s.Require().NoError(err)
+	return messenger
+}
+
+func (s *MessengerShareUrlsSuite) createCommunity() *communities.Community {
+	description := &requests.CreateCommunity{
+		Membership:  protobuf.CommunityPermissions_NO_MEMBERSHIP,
+		Name:        "status",
+		Color:       "#ffffff",
+		Description: "status community description",
+	}
+
+	// Create an community chat
+	response, err := s.m.CreateCommunity(description, false)
+	s.Require().NoError(err)
+	s.Require().NotNil(response)
+	return response.Communities()[0]
+}
+
+func (s *MessengerShareUrlsSuite) createCommunityWithChannel() (*communities.Community, *protobuf.CommunityChat, string) {
+	community := s.createCommunity()
+
+	chat := &protobuf.CommunityChat{
+		Permissions: &protobuf.CommunityPermissions{
+			Access: protobuf.CommunityPermissions_NO_MEMBERSHIP,
+		},
+		Identity: &protobuf.ChatIdentity{
+			DisplayName: "status-core",
+			Emoji:       "😎",
+			Description: "status-core community chat",
+		},
+	}
+	response, err := s.m.CreateCommunityChat(community.ID(), chat)
+	s.Require().NoError(err)
+	s.Require().NotNil(response)
+	s.Require().Len(response.Communities(), 1)
+	s.Require().Len(response.Chats(), 1)
+
+	community = response.Communities()[0]
+	s.Require().Len(community.Chats(), 1)
+
+	var channelId string
+	var channel *protobuf.CommunityChat
+
+	for key, value := range community.Chats() {
+		channelId = key
+		channel = value
+		break
+	}
+	s.Require().NotNil(channel)
+	return community, channel, channelId
+}
+
+func (s *MessengerShareUrlsSuite) TestDecodeEncodeDataURL() {
+	testData := []byte("test data 123")
+
+	encodedData, err := urls.EncodeDataURL(testData)
+	s.Require().NoError(err)
+
+	decodedData, err := urls.DecodeDataURL(encodedData)
+	s.Require().NoError(err)
+	s.Require().Equal(testData, decodedData)
+}
```

Comment:
@borismelnik @MishkaRogachev I have fix for this code test

[fix.patch](https://github.com/status-im/status-go/files/11895490/fix.patch)

You can download it and run `git apply ./fix.patch` it will apply the fix

```patch
diff --git a/protocol/messenger_share_urls_test.go b/protocol/messenger_share_urls_test.go
index 129c6a4f8..b8888dd03 100644
--- a/protocol/messenger_share_urls_test.go
+++ b/protocol/messenger_share_urls_test.go
@@ -112,7 +112,7 @@ func (s *MessengerShareUrlsSuite) createCommunityWithChannel() (*communities.Com
 }
 
 func (s *MessengerShareUrlsSuite) TestDecodeEncodeDataURL() {
-	testData := []byte("test data 123")
+	testData := []byte("test data 123test data 123test data 123test data 123test data 123")
 
 	encodedData, err := urls.EncodeDataURL(testData)
 	s.Require().NoError(err)
diff --git a/protocol/urls/utils.go b/protocol/urls/utils.go
index 00ba40942..4a5bbb0a4 100644
--- a/protocol/urls/utils.go
+++ b/protocol/urls/utils.go
@@ -61,7 +61,11 @@ func encodeDataURL(data []byte) (string, error) {
 	bb := bytes.NewBuffer([]byte{})
 	writer := brotli.NewWriter(bb)
 	_, err := writer.Write(data)
-	writer.Close()
+	if err != nil {
+		return "", err
+	}
+
+	err = writer.Close()
 	if err != nil {
 		return "", err
 	}
@@ -71,17 +75,17 @@ func encodeDataURL(data []byte) (string, error) {
 
 func decodeDataURL(data string) ([]byte, error) {
 	decoded, err := base64.URLEncoding.DecodeString(data)
-
 	if err != nil {
 		return nil, err
 	}
-	var output []byte
+
+	output := make([]byte, 2^12)
 	bb := bytes.NewBuffer(decoded)
 	reader := brotli.NewReader(bb)
-	_, err = reader.Read(output)
+	n, err := reader.Read(output)
 	if err != nil {
 		return nil, err
 	}
 
-	return output, nil
+	return output[:n], nil
 }
```

Reply:
Hang on there is something not right

Reply:
Ah I didn't reply in the right place https://github.com/status-im/status-go/pull/3665#issuecomment-1611519219

Reply:
Thanks, @Samyoul!
I tried to apply path, but in the test data seems to be cropped:
```
--- Expected
                                +++ Actual
                                @@ -1,7 +1,3 @@
                                -([]uint8) (len=65) {
                                - 00000000  74 65 73 74 20 64 61 74  61 20 31 32 33 74 65 73  |test data 123tes|
                                - 00000010  74 20 64 61 74 61 20 31  32 33 74 65 73 74 20 64  |t data 123test d|
                                - 00000020  61 74 61 20 31 32 33 74  65 73 74 20 64 61 74 61  |ata 123test data|
                                - 00000030  20 31 32 33 74 65 73 74  20 64 61 74 61 20 31 32  | 123test data 12|
                                - 00000040  33                                                |3|
                                +([]uint8) (len=14) {
                                + 00000000  74 65 73 74 20 64 61 74  61 20 31 32 33 74        |test data 123t|
                                 }
```